### PR TITLE
Have smem collector add environ variables as dimensions

### PR DESCRIPTION
**Pull request by avadhutp**
Ticket: https://jira.yelpcorp.com/browse/PERF-1029
Corresponding config change is already in: https://reviewboard.yelpcorp.com/r/176601/

### CODE REVIEW

File: src/fullerite/collector/smem.go [hunk 1]
```diff
@@ -150,21 +192,51 @@ func (s *SmemStats) getCustomDimensions(pid int) map[string]string {
...
+	}
+
+	return s.parseSmemLines(string(out))
+}
+
+func (s *SmemStats) getEnviron(pid int) string {
+	cmdLine := []string{
+		"/usr/bin/sudo",
+		"-u", s.user,
+		"/bin/cat",
```
Comment (poros) [inline src/fullerite/collector/smem.go]: Could you just read the file instead of spawning another subshell or is there a permission problem?

Comment (avadhutp) [inline src/fullerite/collector/smem.go]: Yep—it is permission problems. So can't just read it. Therefore, there's an accompanying puppet change with it.

File: src/fullerite/collector/smem.go [hunk 2]
```diff
@@ -126,11 +133,46 @@ func (s *SmemStats) Collect() {
...
+	}
+
+	environ := s.getEnviron(pid)
+
+	if environ == "" {
+		return dims
+	}
+
+	for dimName, envVar := range s.dimensionsFromEnv {
+		pattern, _ := regexp.Compile(fmt.Sprintf(`(%s=)(.*?)(\000)`, envVar))
```
Comment (drolando) [inline src/fullerite/collector/smem.go]: 2 things:
- will it work if the env variable is at the end of the string? Is the last element '\0' terminated?
- can you just write it as (%s)=(.*?)(\000)? I just spent 5 mins searching in the regex syntax manual what that = after %s meant.

File: src/fullerite/collector/smem.go [hunk 3]
```diff
@@ -126,11 +133,46 @@ func (s *SmemStats) Collect() {
...
+	dims := getEnvDimensions(s, pid)
+
+	for k, v := range getCmdLineDimensions(s, pid) {
+		dims[k] = v
+	}
+
+	return dims
+}
+
 func (s *SmemStats) readCmdline(fileName string) ([]byte, error) {
```
Comment (drolando) [inline src/fullerite/collector/smem.go]: Should we use "/usr/bin/sudo -u <user> /bin/cat /proc/<pid>/cmdline" like for environ to keep the code simpler and not read each file in a different way?

File: src/fullerite/collector/smem.go [hunk 4]
```diff
@@ -126,11 +133,46 @@ func (s *SmemStats) Collect() {
 	}
 }
 
+func (s *SmemStats) getCustomDimensions(pid int) map[string]string {
+	dims := getEnvDimensions(s, pid)
+
+	for k, v := range getCmdLineDimensions(s, pid) {
```
Comment (drolando) [inline src/fullerite/collector/smem.go]: :( there doesn't seem to be a way to merge 2 maps in golang, so this is probably the only approach

File: src/fullerite/collector/smem.go (hunk 2 above)
Comment (avadhutp) [inline src/fullerite/collector/smem.go]: - even if it is at the end of the string, it is delimited with a `\000`
- fixed

File: src/fullerite/collector/smem.go (hunk 3 above)
Comment (avadhutp) [inline src/fullerite/collector/smem.go]: Done

**Comment by avadhutp** [comment]
@poros @dichiarafrancesco @drolando  Can I merge this in?

**Pull request merged**